GCC 12 でのコンパイルエラー・警告対応

GCC 12 でのコンパイルを試したところ、以下のエラーと警告が出力されるのを確認（GCC 11 までは問題なし）。
エラーはもちろん、警告もなくコンパイルできるよう対応する。

```sh
Making all in src
make[1]: ディレクトリ '/home/habu/hengband/src' に入ります
make  all-am
make[2]: ディレクトリ '/home/habu/hengband/src' に入ります
depbase=`echo core/scores.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
./gcc-wrap ccache g++-12 -std=gnu++17 -DHAVE_CONFIG_H -I. -I../src   -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/x86_64-linux-gnu -include ../src/stdafx.h  -g -Wall -Wextra -MT core/sc
ores.o -MD -MP -MF $depbase.Tpo -c -o core/scores.o core/scores.cpp &&\                                                                                                                                                         mv -f $depbase.Tpo $depbase.Po
depbase=`echo main/info-initializer.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
./gcc-wrap ccache g++-12 -std=gnu++17 -DHAVE_CONFIG_H -I. -I../src   -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/x86_64-linux-gnu -include ../src/stdafx.h  -g -Wall -Wextra -MT main/in
fo-initializer.o -MD -MP -MF $depbase.Tpo -c -o main/info-initializer.o main/info-initializer.cpp &&\                                                                                                                           mv -f $depbase.Tpo $depbase.Po
core/scores.cpp: In function ‘errr top_twenty(PlayerType*)’:
core/scores.cpp:228:33: warning: ‘%3d’ directive writing between 3 and 5 bytes into a region of size 4 [-Wformat-overflow=]
  228 |     sprintf(the_score.cur_lev, "%3d", std::min<ushort>(player_ptr->lev, 999));
      |                                 ^~~
core/scores.cpp:228:32: note: directive argument in the range [0, 65535]
  228 |     sprintf(the_score.cur_lev, "%3d", std::min<ushort>(player_ptr->lev, 999));
      |                                ^~~~~
core/scores.cpp:228:12: note: ‘sprintf’ output between 4 and 6 bytes into a destination of size 4
  228 |     sprintf(the_score.cur_lev, "%3d", std::min<ushort>(player_ptr->lev, 999));
      |     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
core/scores.cpp:230:33: warning: ‘%3d’ directive writing between 3 and 5 bytes into a region of size 4 [-Wformat-overflow=]
  230 |     sprintf(the_score.max_lev, "%3d", std::min<ushort>(player_ptr->max_plv, 999));
      |                                 ^~~
core/scores.cpp:230:32: note: directive argument in the range [0, 65535]
  230 |     sprintf(the_score.max_lev, "%3d", std::min<ushort>(player_ptr->max_plv, 999));
      |                                ^~~~~
core/scores.cpp:230:12: note: ‘sprintf’ output between 4 and 6 bytes into a destination of size 4
  230 |     sprintf(the_score.max_lev, "%3d", std::min<ushort>(player_ptr->max_plv, 999));
      |     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
core/scores.cpp: In function ‘errr predict_score(PlayerType*)’:
core/scores.cpp:333:33: warning: ‘%3d’ directive writing between 3 and 5 bytes into a region of size 4 [-Wformat-overflow=]
  333 |     sprintf(the_score.cur_lev, "%3d", std::min<ushort>(player_ptr->lev, 999));
      |                                 ^~~
core/scores.cpp:333:32: note: directive argument in the range [0, 65535]
  333 |     sprintf(the_score.cur_lev, "%3d", std::min<ushort>(player_ptr->lev, 999));
      |                                ^~~~~
core/scores.cpp:333:12: note: ‘sprintf’ output between 4 and 6 bytes into a destination of size 4
  333 |     sprintf(the_score.cur_lev, "%3d", std::min<ushort>(player_ptr->lev, 999));
      |     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
core/scores.cpp:335:33: warning: ‘%3d’ directive writing between 3 and 5 bytes into a region of size 4 [-Wformat-overflow=]
  335 |     sprintf(the_score.max_lev, "%3d", std::min<ushort>(player_ptr->max_plv, 999));
      |                                 ^~~
core/scores.cpp:335:32: note: directive argument in the range [0, 65535]
  335 |     sprintf(the_score.max_lev, "%3d", std::min<ushort>(player_ptr->max_plv, 999));
      |                                ^~~~~
core/scores.cpp:335:12: note: ‘sprintf’ output between 4 and 6 bytes into a destination of size 4
  335 |     sprintf(the_score.max_lev, "%3d", std::min<ushort>(player_ptr->max_plv, 999));
      |     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
main/info-initializer.cpp: In instantiation of ‘errr init_info(std::string_view, angband_header&, InfoType&, Parser, {anonymous}::Retoucher) [with InfoType = std::vector<TerrainType>; errr = int; std::string_view = std::basi
c_string_view<char>; Parser = std::function<int(std::basic_string_view<char>, angband_header*)>; {anonymous}::Retoucher = void (*)(angband_header*)]’:                                                                          main/info-initializer.cpp:204:21:   required from here
main/info-initializer.cpp:105:20: error: no matching function for call to ‘std::vector<TerrainType>::assign(uint16_t&, <brace-enclosed initializer list>)’
  105 |         info.assign(head.info_num, {});
      |         ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/12/vector:64,
                 from /usr/include/c++/12/functional:62,
                 from ./stdafx.h:5,
                 from ./stdafx.cpp:1:
/usr/include/c++/12/bits/stl_vector.h:823:9: note: candidate: ‘template<class _InputIterator, class> void std::vector<_Tp, _Alloc>::assign(_InputIterator, _InputIterator) [with <template-parameter-2-2> = _InputIterator; _Tp 
= TerrainType; _Alloc = std::allocator<TerrainType>]’                                                                                                                                                                             823 |         assign(_InputIterator __first, _InputIterator __last)
      |         ^~~~~~
/usr/include/c++/12/bits/stl_vector.h:823:9: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/12/bits/stl_algobase.h:65,
                 from /usr/include/c++/12/algorithm:60,
                 from ./stdafx.h:2:
/usr/include/c++/12/bits/stl_iterator_base_types.h: In substitution of ‘template<class _InIter> using _RequireInputIter = std::__enable_if_t<std::is_convertible<typename std::iterator_traits< <template-parameter-1-1> >::iter
ator_category, std::input_iterator_tag>::value> [with _InIter = short unsigned int]’:                                                                                                                                           /usr/include/c++/12/bits/stl_vector.h:820:9:   required from ‘errr init_info(std::string_view, angband_header&, InfoType&, Parser, {anonymous}::Retoucher) [with InfoType = std::vector<TerrainType>; errr = int; std::string_vi
ew = std::basic_string_view<char>; Parser = std::function<int(std::basic_string_view<char>, angband_header*)>; {anonymous}::Retoucher = void (*)(angband_header*)]’                                                             main/info-initializer.cpp:204:21:   required from here
/usr/include/c++/12/bits/stl_iterator_base_types.h:249:11: error: no type named ‘iterator_category’ in ‘struct std::iterator_traits<short unsigned int>’
  249 |     using _RequireInputIter =
      |           ^~~~~~~~~~~~~~~~~
main/info-initializer.cpp: In instantiation of ‘errr init_info(std::string_view, angband_header&, InfoType&, Parser, {anonymous}::Retoucher) [with InfoType = std::vector<TerrainType>; errr = int; std::string_view = std::basi
c_string_view<char>; Parser = std::function<int(std::basic_string_view<char>, angband_header*)>; {anonymous}::Retoucher = void (*)(angband_header*)]’:                                                                          main/info-initializer.cpp:204:21:   required from here
/usr/include/c++/12/bits/stl_vector.h:803:7: note: candidate: ‘void std::vector<_Tp, _Alloc>::assign(size_type, const value_type&) [with _Tp = TerrainType; _Alloc = std::allocator<TerrainType>; size_type = long unsigned int;
 value_type = TerrainType]’                                                                                                                                                                                                       803 |       assign(size_type __n, const value_type& __val)
      |       ^~~~~~
/usr/include/c++/12/bits/stl_vector.h:803:47: note:   no known conversion for argument 2 from ‘<brace-enclosed initializer list>’ to ‘const std::vector<TerrainType>::value_type&’ {aka ‘const TerrainType&’}
  803 |       assign(size_type __n, const value_type& __val)
      |                             ~~~~~~~~~~~~~~~~~~^~~~~
/usr/include/c++/12/bits/stl_vector.h:850:7: note: candidate: ‘void std::vector<_Tp, _Alloc>::assign(std::initializer_list<_Tp>) [with _Tp = TerrainType; _Alloc = std::allocator<TerrainType>]’
  850 |       assign(initializer_list<value_type> __l)
      |       ^~~~~~
/usr/include/c++/12/bits/stl_vector.h:850:7: note:   candidate expects 1 argument, 2 provided
make[2]: *** [Makefile:6009: main/info-initializer.o] エラー 1
make[2]: ディレクトリ '/home/habu/hengband/src' から出ます
make[1]: *** [Makefile:2739: all] エラー 2
make[1]: ディレクトリ '/home/habu/hengband/src' から出ます
make: *** [Makefile:385: all-recursive] エラー 1
```

--- a/src/main/info-initializer.cpp
+++ b/src/main/info-initializer.cpp
@@ -102,7 +102,8 @@ static errr init_info(std::string_view filename, angband_header &head, InfoType 
 
     constexpr auto info_is_vector = is_vector_v<InfoType>;
     if constexpr (info_is_vector) {
-        info.assign(head.info_num, {});
+        using value_type = typename InfoType::value_type;
+        info.assign(head.info_num, value_type{});
     }
 
     const auto err = init_info_txt(fp, buf, &head, parser);

